extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java

```diff
@@ -297,29 +297,55 @@ void registerClasses(
-            List<JaxbClassesToBeBoundBuildItem> classesToBeBoundBuildItems,
-            JaxbContextConfigRecorder jaxbContextConfig) {
-        Set<String> classNamesToBeBound = new HashSet<>();
-        for (JaxbClassesToBeBoundBuildItem classesToBeBoundBuildItem : classesToBeBoundBuildItems) {
-            classNamesToBeBound.addAll(classesToBeBoundBuildItem.getClasses());
-        }
+    FilteredJaxbClassesToBeBoundBuildItem filterBoundClasses(
+            JaxbConfig config,
+            List<JaxbClassesToBeBoundBuildItem> classesToBeBoundBuildItems) {
+
+        FilteredJaxbClassesToBeBoundBuildItem.Builder builder = FilteredJaxbClassesToBeBoundBuildItem.builder();
+        classesToBeBoundBuildItems.stream()
+                .map(JaxbClassesToBeBoundBuildItem::getClasses)
+                .forEach(builder::classNames);
 
         // remove classes that have been excluded by users
         if (config.excludeClasses.isPresent()) {
-            classNamesToBeBound.removeAll(config.excludeClasses.get());
+            builder.classNameExcludes(config.excludeClasses.get());
         }
+        return builder.build();
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void setupJaxbContextConfig(
+            FilteredJaxbClassesToBeBoundBuildItem filteredClassesToBeBound,
+            JaxbContextConfigRecorder jaxbContextConfig) {
+        jaxbContextConfig.addClassesToBeBound(filteredClassesToBeBound.getClasses());
+    }
 
-        // parse class names to class
-        Set<Class<?>> classes = getAllClassesFromClassNames(classNamesToBeBound);
-        if (config.validateJaxbContext) {
-            // validate the context to fail at build time if it's not valid
-            validateContext(classes);
+    @BuildStep
+    void validateDefaultJaxbContext(
+            JaxbConfig config,
+            FilteredJaxbClassesToBeBoundBuildItem filteredClassesToBeBound,
+            SynthesisFinishedBuildItem beanContainerState,
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> errors) {
+
+        final BeanResolver beanResolver = beanContainerState.getBeanResolver();
+        final Set<BeanInfo> beans = beanResolver
+                .resolveBeans(Type.create(DotName.createSimple(JAXBContext.class), org.jboss.jandex.Type.Kind.CLASS));
+        if (config.validateJaxbContext && !beans.isEmpty()) {
+            final BeanInfo bean = beanResolver.resolveAmbiguity(beans);
+            if (bean.isDefaultBean()) {
+                /*
+                 * Validate the default JAXB context at build time and fail early.
+                 * Do this only if the user application actually requires the default JAXBContext bean
+                 */
+                try {
+                    JAXBContext.newInstance(filteredClassesToBeBound.getClasses().toArray(new Class[0]));
+                } catch (JAXBException e) {
+                    throw new IllegalStateException("Failed to configure JAXB context", e);
```

Comment:
Produce a ValidationErrorBuildItem build item instead of throwing this exception.

Reply:
Good idea, let me do that. 
I just copied the old code without too much thinking about it. I had to add the ValidationErrorBuildItem consumer to the method signature to force the build time container to invoke the method.

Reply:
I had to remove the ValidationErrorBuildItem producer from the method signature because of the dependency cycle in resteasy-reactive-jaxb. Now the DeploymentException is thrown directly from validateDefaultJaxbContext() much like it was before this change.

Reply:
ok ok!